ledger/heavy/executor/jetkeeper.go

```diff
@@ -20,143 +20,358 @@ import (
+		". top: ", top, ". prev.PulseNumber: ", prev.PulseNumber)
+
+	if prev.PulseNumber != top {
+		// We should sync pulses sequentially. We can't skip.
+		return nil
+	}
+
+	for jk.checkPulseConsistency(ctx, pn) {
+		err := jk.updateSyncPulse(pn)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update consistent pulse")
+		}
+		logger.Debugf("pulse completed: %d", pn)
+
+		next, err := jk.pulses.Forwards(ctx, pn, 1)
+		if err == pulse.ErrNotFound {
+			logger.Info("propagateConsistency. No next pulse. Stop propagating")
+			return nil
+		}
+		if err != nil {
+			return errors.Wrapf(err, "failed to get next pulse for %d", pn)
+		}
+		pn = next.PulseNumber
+	}
+
 	return nil
 }
 
 func (jk *dbJetKeeper) TopSyncPulse() insolar.PulseNumber {
 	jk.RLock()
 	defer jk.RUnlock()
 
-	it := jk.db.NewIterator(syncPulseKey(0xFFFFFFFF), true)
-	defer it.Close()
-	if it.Next() {
-		return insolar.NewPulseNumber(it.Key()[1:])
+	return jk.topSyncPulse()
+}
+
+func (jk *dbJetKeeper) topSyncPulse() insolar.PulseNumber {
+	val, err := jk.db.Get(syncPulseKey{})
+	if err != nil {
+		return insolar.GenesisPulse.PulseNumber
 	}
-	return insolar.GenesisPulse.PulseNumber
+	return insolar.NewPulseNumber(val)
 }
 
-func (jk *dbJetKeeper) add(pulse insolar.PulseNumber, id insolar.JetID) error {
+func (jk *dbJetKeeper) updateHot(ctx context.Context, pulse insolar.PulseNumber, id insolar.JetID, split bool) error {
+	logger := inslogger.FromContext(ctx)
 	jets, err := jk.get(pulse)
-	if err != nil {
-		jets = []insolar.JetID{}
+	if err != nil && err != store.ErrNotFound {
+		return errors.Wrapf(err, "updateHot. can't get pulse: %d", pulse)
+	}
+	var exists bool
+	parentID := id
+	if split {
+		parentID = jet.Parent(id)
+	}
+	for i := range jets {
+		if jets[i].JetID.Equal(parentID) {
+			exists = true
+			err := jets[i].addHot(id, split)
+			if err != nil {
+				return errors.Wrap(err, "can't addHot")
+			}
+			logger.Debug("updateHot. update existing. pulse: ", pulse, ", Jet:", id.DebugString(), ", split: ", split)
+			break
+		}
+	}
+	if !exists {
+		newInfo := jetInfo{}
+		err := newInfo.addHot(id, split)
+		if err != nil {
+			return errors.Wrap(err, "can't addHot ( not existing )")
+		}
+		jets = append(jets, newInfo)
+		logger.Debug("updateHot. not existing. pulse: ", pulse, ". Jet:", id.DebugString(), ", split: ", split)
 	}
-	jets = append(jets, id)
 	return jk.set(pulse, jets)
 }
 
-func (jk *dbJetKeeper) checkPulseConsistency(ctx context.Context, pulse insolar.PulseNumber) error {
-	logger := inslogger.FromContext(ctx).WithFields(map[string]interface{}{
-		"pulse": pulse,
-	})
+func (jk *dbJetKeeper) updateDrop(ctx context.Context, pulse insolar.PulseNumber, id insolar.JetID, split bool) error {
+	logger := inslogger.FromContext(ctx)
+	jets, err := jk.get(pulse)
+	if err != nil && err != store.ErrNotFound {
+		return errors.Wrapf(err, "updateDrop. can't get pulse: %d", pulse)
+	}
+	var exists bool
```

Comment:
Let's make `getForJet(pn, jet)` function that creates absent keys. I think, it will make these functions a bit cleaner.